assign clusters to clusterclaims

staebler opened the pull request:
A ClusterClaim resource is created by a user when they want to claim a cluster from a ClusterPool. The ClusterClaim must be in the same namespace as the ClusterPool. When a cluster has been assigned to the ClusterClaim, the namespace field in the status will be filled out with the namespace where the ClusterDeployment for the cluster resides, and the NotReady condition will be set to False.

When reconciling a ClusterPool, the clusterpool controller will check for ClusterClaims that reference the ClusterPool and that do not have a cluster assigned. If there are ready clusters, the controller will set the `.status.namespace` field to the namespace of one of the ready clusters. If there are not enough ready clusters to fulfill all claims, the controller will prioritize the oldest claims. If a claim is not assigned a cluster, the NotReady condition will be set to True on the claim with a reason of NoClusters.

The clusterclaim controller has been added to handle reconciling ClusterClaims that have been assigned a cluster. For a claim with an assignment, the controller will check the ClusterDeployment to ensure that the cluster has not been assigned to a different claim. If it has, the the `.status.namespace` field of the ClusterClaim will be cleared. If the ClusterDeployment has not been associated with any claim, then the controller will update the ClusterDeployment to associated it with the ClusterClaim, completing the cluster assignment. When the ClusterClaim is deleted, the controller will delete the ClusterDeployment assigned  to the claim.

https://issues.redhat.com/browse/CO-1012

staebler commented:
/hold

This builds on https://github.com/openshift/hive/pull/1060.

staebler commented:
#1060 has merged.

/hold cance;l

staebler commented:
#1060 has merged.

/hold cancel

dgoodwin commented on pkg/controller/clusterpool/clusterpool_controller.go:
```diff
@@ -513,3 +536,71 @@ func (r *ReconcileClusterPool) createCloudBuilder(pool *hivev1.ClusterPool, logg
+				hivev1.ClusterClaimNotReadyCondition,
+				corev1.ConditionTrue,
+				"ClusterAssigned",
+				"Cluster assigned to claim",
```
We're still NotReady=true here, could we expand the message a little to indicate we're waiting for the link to be completed, maybe "Cluster identified for claim, awaiting assignment" or something like that.

dgoodwin commented on pkg/apis/hive/v1/clusterclaim_types.go:
```diff
@@ -0,0 +1,83 @@
+// ClusterClaimStatus defines the observed state of ClusterClaim
+type ClusterClaimStatus struct {
+	// Namespace is the namespace containing the ClusterDeployment of the claimed cluster.
+	Namespace string `json:"namespace,omitempty"`
```
Maybe.. see below.

staebler replied:
I like moving it to the Spec. I'll do that.

staebler commented on pkg/controller/clusterclaim/clusterclaim_controller.go:
```diff
@@ -0,0 +1,234 @@
+// Reconcile reconciles a ClusterClaim.
+func (r *ReconcileClusterClaim) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	start := time.Now()
+	logger := r.logger.WithField("clusterClaim", request.Name)
```
It already is. See https://github.com/openshift/hive/blob/d7b5fab774e8ac330a4876b987f4db1fbaad4406/pkg/controller/clusterclaim/clusterclaim_controller.go#L112.

staebler replied:
Yep, that should be the following.
```suggestion
	logger := r.logger.WithField("clusterClaim", request.NamespacedName)
```

staebler commented on pkg/apis/hive/v1/clusterclaim_types.go:
```diff
@@ -0,0 +1,83 @@
+	LastProbeTime metav1.Time `json:"lastProbeTime,omitempty"`
+	// LastTransitionTime is the last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
```
We can add a field if it is needed for metrics when we add metrics.

staebler commented on pkg/controller/clusterclaim/clusterclaim_controller.go:
```diff
@@ -0,0 +1,234 @@
+		case err != nil:
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "error get ClusterDeployment")
+			return reconcile.Result{}, err
+		case cd.Spec.ClusterPoolRef == nil, cd.Spec.ClusterPoolRef.Namespace != claim.Namespace, cd.Spec.ClusterPoolRef.ClaimName != claim.Name:
```
If you prefer. I personally don't think that it warrants a warning level. I expect this to happen when two claims are created close in time to each other. There is a chance that the clusterpool controller will assign the same cluster to both of the claims. Whichever claim wins the race in the clusterclaim controller ultimately gets the cluster.

dgoodwin replied:
No that's fine if it's common and expected.

staebler commented on pkg/controller/clusterclaim/clusterclaim_controller.go:
```diff
@@ -0,0 +1,234 @@
+		"ClusterDeleted",
+		"Assigned cluster has been deleted",
+		controllerutils.UpdateConditionIfReasonOrMessageChange,
+	)
```
Yes, Hive will do nothing. The condition is just for informationally purposes.

staebler commented on pkg/controller/clusterclaim/clusterclaim_controller.go:
```diff
@@ -0,0 +1,234 @@
+}
+
+func (r *ReconcileClusterClaim) reconcileForNewAssignment(claim *hivev1.ClusterClaim, cd *hivev1.ClusterDeployment, logger log.FieldLogger) (reconcile.Result, error) {
+	logger.Info("cluster assigned to claim")
```
What context are you looking for. The logger already has a field for the cluster name.

dgoodwin replied:
Ah I went looking but couldn't see it, if there's context for the cd name then I'm good.

staebler commented:
Addressed code review feedback.
* Moved the `.status.namespace` field of `ClusterClaim` to `.spec.namespace`. This preserves the cluster assignment if the `ClusterClaim` loses its status, say due to a restore from backup.
* Changed terminology around cluster assignment. When the clusterpool controller sets the `.spec.namespace` field of a `ClusterClaim`, that is "assigning" the cluster to the `ClusterClaim`. When the clusterclaim controller sets the `.spec.clusterPoolRef.claimName` field of a `ClusterDeployment`, that is "claiming" the cluster.

dgoodwin commented:
/lgtm
